File: src/python/_0779_K-th_Symbol_in_Grammar.py
class Solution:
    def kthGrammar(self, n: int, k: int) -> int:
        def dfs(i, j):
            if i == 0:
                return 0

            if i == 1:
                return j

            divisor = 2 ** (i - 1)

            if divisor > j:
                return dfs(i - 1, j)
            elif dfs(i - 1, j % divisor) == 0:
                return 1
            else:
                return 0

        return dfs(n - 1, k - 1)

File: src/_0779_K-th_Symbol_in_Grammar.java
class Solution {
    private int dfs(int n, int k) {
        if (n == 0)
            return 0;

        if (n == 1)
            return k;

        int power = (int) Math.pow(2, n - 1);

        if (power > k)
            return dfs(n - 1, k);
        else if (dfs(n - 1, k % power) == 1)
            return 0;
        else
            return 1;
    }

    public int kthGrammar(int n, int k) {
        return dfs(n - 1, k - 1);
    }
}
